File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.py
class ProductOfNumbers:

    def __init__(self):
        self.products = []
        self.product = 1

        

    def add(self, num: int) -> None:
        if num:
            self.product *= num
            self.products.append(self.product)
        else:
            self.product = 1
            self.products = []
        

        

    def getProduct(self, k: int) -> int:
        if len(self.products) < k:
            return 0
        elif k == len(self.products):
            return self.product
        else:
            return (self.products[-1] // self.products[-k-1]  )

        


# Your ProductOfNumbers object will be instantiated and called as such:
# obj = ProductOfNumbers()
# obj.add(num)
# param_2 = obj.getProduct(k)

File: 1477-product-of-the-last-k-numbers/product-of-the-last-k-numbers.java
class ProductOfNumbers {
    private List<Integer> products;
    private int product;

    public ProductOfNumbers() 
    {
        products = new ArrayList<>();
        product = 1 ;
        
    }
    
    public void add(int num) 
    {
        if (num != 0)
        {
            product *= num ;
            products.add(product) ;
        }
        else
        {
            products.clear() ;
            product = 1 ;

        }
        
    }
    
    public int getProduct(int k) 
    {
        if (products.size() < k)
        {
            return 0 ;
        }
        else if (products.size() == k)
        {
            return product ;
        }
        else
        {
            return products.get(products.size() - 1) / products.get(products.size() - k - 1);
        }
        
    }
}

/**
 * Your ProductOfNumbers object will be instantiated and called as such:
 * ProductOfNumbers obj = new ProductOfNumbers();
 * obj.add(num);
 * int param_2 = obj.getProduct(k);
 */